Handle bold/strong nested inside italics/em

cddude229 opened the pull request:
While trying out this library, my initial input ran up against some edge cases.  (It handled everything else perfectly though, which is awesome.). I decided to take a stab at fixing the bugs, which is this PR.  I've broken down every test case addition + related fix into its own commit for easy review.

Here's the summary of changes in input/output:
| Input | Output in production | Output after this PR |
| ------|---------------|-------------------|
| `*italics **and bold** end*` | `<p>*italics <strong>and bold</strong> end*</p>` | `<p><em>italics <strong>and bold</strong> end</em></p>` |
| `*italics **and bold***` | `<p>*italics <strong>and bold</strong>*</p>` | `<p><em>italics <strong>and bold</strong></em></p>` |
| `***bold** and italics*` | `<p>*<strong>bold</strong> and italics*</p>` | `<p><em><strong>bold</strong> and italics</em></p>` |
| `*start **bold** and italics*` | `<p>*start <strong>bold</strong> and italics*</p>` | `<p><em>start <strong>bold</strong> and italics</em></p>` |
| `*improper **nesting* is** bad` | `<p>*improper <strong>nesting* is</strong> bad</p>` | `<p><em>improper **nesting</em> is** bad</p>` |

miekg commented:
I don't see how the testcase makes the case for this PR? 

Don't understand the random `i +=2`. Comments also didn't change even though the code around it changed very much

kjk commented:
@miekg It does seem to fix the following:
In current code:
`*italics **and bold** end*`
=>
`<p>*italics <strong>and bold</strong> end*</p>`
See https://tools.arslexis.io/goplayground/#iPKnCL5art6

Per babelmark, most other return: 
`<p><em>italics <strong>and bold </strong> end</em></p>`
see: https://babelmark.github.io/?text=*italics+**and+bold**+end*

The case `*improper **nesting* is** bad` gets closer to what majority does but not fully (see https://babelmark.github.io/?text=*improper+**nesting*+is**+bad)

cddude229 commented:
> I don't see how the testcase makes the case for this PR?
> 
> Don't understand the random `i +=2`. Comments also didn't change even though the code around it changed very much

Apologies - I got lazy at the end of the work day and should have provided more context.  I've updated the PR description and split the main commit into two (to better highlight what each change does) in hopes it better communicates the goals.

I also realized another edge case and added a test for that + fixed it.  This reverted the optimization case I was talking about as the bug was there instead.

Lemme know if you need more information!

cddude229 commented on parser/inline.go:
```diff
@@ -1201,7 +1201,7 @@ func helperEmphasis(p *Parser, data []byte, c byte) (int, ast.Node) {
 
 		if i+1 < len(data) && data[i+1] == c {
-			i++
+			i += 2
```
`i++` guarantees the next character checked is `c`.  This means `helperFindEmphChar` above returns a guaranteed `0` for length, which then short-circuits to the `return 0, nil` case.  We choose to skip the next `c` character here by incrementing by 2 instead.  (This will expose a separate bug if we're actually in the triple `c` case)

cddude229 commented on parser/inline.go:
```diff
@@ -1192,9 +1192,6 @@ func helperEmphasis(p *Parser, data []byte, c byte) (int, ast.Node) {
 
 	for i < len(data) {
 		length := helperFindEmphChar(data[i:], c)
-		if length == 0 {
```
Building on last commit, in the triple `c` case, we'd still have short-circuited here - but we actually want to claim this `c` as our own.  So we do the rest of the validity checks first, and then do this check at the end to exit the loop

cddude229 commented:
@miekg Just following up on this, as I think I addressed all your concerns.  Let me know your thoughts and if you need me to do anything else!

kjk commented:
Thanks, I pushed it in https://github.com/gomarkdown/markdown/commit/77f4768f716dcbd724917f1bade5f8c07e8a8cc1

Had to resolve a conflict and add a fix for infinite loop from #311 so couldn't just merge it
